transports/quic/src/hole_punching.rs

```diff
@@ -1,47 +1,42 @@
-use std::{net::SocketAddr, time::Duration};
+use crate::{Error, Provider};
 
 use futures::future::Either;
+
 use rand::{distributions, Rng};
 
-use crate::{
-    endpoint::{self, ToEndpoint},
-    Error, Provider,
+use std::{
+    io,
+    net::{SocketAddr, UdpSocket},
+    time::Duration,
 };
 
 pub(crate) async fn hole_puncher<P: Provider>(
-    endpoint_channel: endpoint::Channel,
+    socket: UdpSocket,
     remote_addr: SocketAddr,
     timeout_duration: Duration,
 ) -> Error {
-    let punch_holes_future = punch_holes::<P>(endpoint_channel, remote_addr);
+    let punch_holes_future = punch_holes::<P>(socket, remote_addr);
     futures::pin_mut!(punch_holes_future);
     match futures::future::select(P::sleep(timeout_duration), punch_holes_future).await {
         Either::Left(_) => Error::HandshakeTimedOut,
         Either::Right((hole_punch_err, _)) => hole_punch_err,
     }
 }
 
-async fn punch_holes<P: Provider>(
-    mut endpoint_channel: endpoint::Channel,
-    remote_addr: SocketAddr,
-) -> Error {
+async fn punch_holes<P: Provider>(socket: UdpSocket, remote_addr: SocketAddr) -> Error {
     loop {
         let sleep_duration = Duration::from_millis(rand::thread_rng().gen_range(10..=200));
         P::sleep(sleep_duration).await;
 
-        let random_udp_packet = ToEndpoint::SendUdpPacket(quinn_proto::Transmit {
-            destination: remote_addr,
-            ecn: None,
-            contents: rand::thread_rng()
-                .sample_iter(distributions::Standard)
-                .take(64)
-                .collect(),
-            segment_size: None,
-            src_ip: None,
-        });
+        let contents: Vec<u8> = rand::thread_rng()
+            .sample_iter(distributions::Standard)
+            .take(64)
+            .collect();
 
-        if endpoint_channel.send(random_udp_packet).await.is_err() {
-            return Error::EndpointDriverCrashed;
+        if let Err(e) = socket.send_to(&contents, remote_addr) {
+            if !matches!(e.kind(), io::ErrorKind::WouldBlock) {
```

Comment:
So this isn't a busy-loop because we have our timeout at the top. Kind of a hack but I guess it works :)

Reply:
Given that we have access to `Provider` here, how about using the corresponding async-std or tokio `send` methods and `await`ing them through the `Provider` trait?

With the current approach, in case the underlying UDP socket can not handle the hole UDP datagram at once, we will never send a single one, right?

Reply:
@kpp @thomaseizinger https://github.com/kpp/rust-libp2p/pull/28/ applies the above suggestion. Let me know what you think.

Reply:
Looks good but this leaks the underlying tokio/async-std runtime. If that's okay, I don't mind merging it.

Reply:
https://github.com/kpp/rust-libp2p/pull/28 is updated and no longer leaks the types. @kpp https://github.com/kpp/rust-libp2p/pull/28 is ready to merge from my side.